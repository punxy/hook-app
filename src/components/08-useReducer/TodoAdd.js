import React from 'react'
import { useForm } from '../../hook/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [{description}, handleInputChange, reset] = useForm({
        description: ''
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if(description.trim().length <= 1){
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            description: description,
            done: false
        };

        handleAddTodo(newTodo);
        reset();
    } 

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />

            <form onSubmit={handleSubmit}>
                <input 
                    type='text' 
                    name='description'
                    className='form-control'
                    placeholder='aprender'
                    autoComplete='off'
                    onChange={handleInputChange}
                    value={description}
                />

                <button 
                    type='submit' 
                    className='btn btn-primary btn-sm btn-block mt-1'
                    >
                    Agregar
                </button>
            </form>
        </>
    )
}
