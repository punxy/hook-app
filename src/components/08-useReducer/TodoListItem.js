import React from 'react'

export const TodoListItem = ({ todo, handleDeleteTodo, handleToggle, index }) => {
    return (
        <div className='row m-1'>
            <div className='col-6'>
                <p 
                    onClick={() => handleToggle(todo.id)}
                    className={`${ todo.done && 'complete' }`}
                    >{ index + 1 } {todo.description}</p>
            </div>
            <div className='col-6'>
                <button
                    className='btn btn-danger btn-sm'
                    onClick={() => handleDeleteTodo(todo.id)}
                    >
                    borrar
                </button>
            </div>
        </div>
    )
}