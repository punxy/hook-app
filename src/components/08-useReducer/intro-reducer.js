 console.log('Intro useReducer');

 const initialState = [{
    id: 1,
    todo: 'Comprar leche',
    done: false
 }];

 const todoReducer = (state = initialState, action) => {

    if(action?.type === 'ADD_TODO'){
        return [
            ...initialState,
            action.payload
        ]
    }

    return state;
 }

 // 'todos' ser√≠a el nuevo State
 let todos = todoReducer();


 const newTodo = {
    id: 2,
    todo: 'Comprar Pan',
    done: false
 };

 const action = {
     type: 'ADD_TODO',
     payload: newTodo
 }

todos = todoReducer(todos, action);

console.log(todos);