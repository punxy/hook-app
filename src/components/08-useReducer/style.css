body {
    padding: 70px;
}

p {
    margin: 0;
    cursor: pointer;
}

.complete {
    text-decoration: line-through;
}