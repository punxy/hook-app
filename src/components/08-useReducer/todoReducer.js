export const todoReducer = (state = [], action) => {

    switch (action.type) {
        case 'ADD_TODO':
            return [ ...state, action.payload ];
        
        case 'DELETE_TODO':
            return state.filter(todo => todo.id !== action.payload);
        
        case 'TOGGLE_TODO':
            // return state.map(todo => {
            //     if(todo.id === action.payload){
            //         return {
            //             ...todo,
            //             done: !todo.done
            //         } 
            //     }

            //     return todo;
            // });

            return state.map(todo => 
                (todo.id === action.payload) 
                    ? { ...todo, done: !todo.done } 
                    : todo
            )
    
        default:
            return state;
    }
};