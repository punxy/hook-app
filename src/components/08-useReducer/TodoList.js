import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos, handleDeleteTodo, handleToggle}) => {
  return (
      <>
        {
            todos.map( (todo, index) => (
                <TodoListItem 
                    key={todo.id}
                    todo={todo}
                    index={index}
                    handleDeleteTodo={handleDeleteTodo}
                    handleToggle={handleToggle}
                />
            ))
        }
      </>
    
  )
}
