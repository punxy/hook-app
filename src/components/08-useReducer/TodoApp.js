import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import './style.css';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {
    const [todos, dispatchTodo] = useReducer(todoReducer, [], init);
    

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos])

    const handleAddTodo = newTodo => {
        dispatchTodo({
            type: 'ADD_TODO',
            payload: newTodo
        });
    };

    const handleDeleteTodo = (todoId) => {
        const actionDeleteTodo = {
            type: 'DELETE_TODO',
            payload: todoId
        };

        dispatchTodo(actionDeleteTodo);
    };

    const handleToggle = (todoId) => {
        dispatchTodo({
            type: 'TOGGLE_TODO',
            payload: todoId
        });
    }

    return (
        <div>
            <h1>TodoApp <small>({todos.length})</small> </h1>
            <hr />

            <div className='row'>
                <div className='col-7'>
                    <TodoList 
                        todos={todos} 
                        handleDeleteTodo={handleDeleteTodo} 
                        handleToggle={handleToggle}
                    />
                </div>

                <div className='col-5'>
                    <TodoAdd
                        handleAddTodo={handleAddTodo}
                    />
                </div>
            </div>
        </div>
    )
}
