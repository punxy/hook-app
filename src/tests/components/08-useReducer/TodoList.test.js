import React from 'react';
import { shallow } from "enzyme";
import '@testing-library/jest-dom';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Test in TodoList component', () => {
    const handleDeleteTodo = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow( <TodoList 
        todos={demoTodos}
        handleDeleteTodo={handleDeleteTodo}
        handleToggle={handleToggle}
    /> );

    test('should be render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should be render all TODOs correctly', () => {
        expect( wrapper.find('TodoListItem').length ).toBe(demoTodos.length);
        const item = wrapper.find('TodoListItem').at(0);
        expect( item.prop('handleDeleteTodo') ).toEqual( expect.any(Function) );
        expect( item.prop('handleToggle') ).toEqual( expect.any(Function) );
    });
});