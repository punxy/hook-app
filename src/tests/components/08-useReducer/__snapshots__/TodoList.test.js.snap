// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in TodoList component should be render correctly 1`] = `
<Fragment>
  <TodoListItem
    handleDeleteTodo={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "description": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDeleteTodo={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "description": "Aprender Node",
        "done": false,
        "id": 2,
      }
    }
  />
</Fragment>
`;
